### PR TITLE
feat(lanelet2_map_validator): add test codes for existing validators

### PR DESCRIPTION
**Pull request by TaikiYamada4**
## Description

This PR revises the test system in lanelet2_map_validator such like
- Split the test code for each validator
- Added small maps to validate each issue detection.
- Added medium size map to check validation works well with normal maps

I made the medium size map (sample_map.osm) by modifying lanelet2_map.osm from `autoware_test_utils`.
I modified that map so that the current requirements in the `autoware_requirement_set.json` will pass. 

## Related links

None

### 1. Test codes

I tested the test codes through

```
colcon test --event-handlers console_cohesion+ --packages-select autoware_lanelet2_map_validator
```

### 2. Test sample_map.osm

I tested that sample_map.osm passes `vm-02-02`, `vm-04-01` and `vm-05-01` through the following command.
```bash
ros2 run autoware_lanelet2_map_validator autoware_lanelet2_map_validator -p mgrs -m ./install/autoware_lanelet2_map_validator/share/autoware_lanelet2_map_validator/data/map/sample_map.osm -i ./src/autoware_tools/map/autoware_lanelet2_map_validator/autoware_requirement_set.json -o ./
```
The figure below is the console output I got.
Screenshot from 2024-11-15 11-41-50

## Notes for reviewers

None

## Interface changes

None

## Effects on system behavior

Test algorithm changed.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: map/autoware_lanelet2_map_validator/test/src/map_validation_tester.hpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      ament_index_cpp::get_package_share_directory("autoware_lanelet2_map_validator");
+
+    map = lanelet::load(
+      package_share_directory + "/data/map/" + file_name, *projector, &loading_errors);
+
+    EXPECT_NE(map, nullptr);
+  }
+
+  lanelet::LaneletMapPtr map{nullptr};
+  std::vector<std::string> loading_errors;
```
Comment (YamatoAndo) [inline map/autoware_lanelet2_map_validator/test/src/map_validation_tester.hpp:45]:
```suggestion
  std::vector<std::string> loading_errors_;
```

File: map/autoware_lanelet2_map_validator/test/src/test_regulatory_elements_details_for_traffic_lights.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+TEST_F(TestRegulatoryElementDetailsForTrafficLights, MissingRefLine)  // NOLINT for gtest
+{
+  load_target_map("traffic_light/traffic_light_regulatory_element_without_ref_line.osm");
+
+  lanelet::validation::RegulatoryElementsDetailsForTrafficLightsValidator checker;
+  const auto & issues = checker(*map);
+
+  for (const auto & error : loading_errors) {
+    std::cout << error << std::endl;
+  }
```
Comment (YamatoAndo) [inline map/autoware_lanelet2_map_validator/test/src/test_regulatory_elements_details_for_traffic_lights.cpp:102]:
```suggestion
```

File: map/autoware_lanelet2_map_validator/test/src/test_regulatory_elements_details_for_crosswalks.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+  lanelet::validation::RegulatoryElementsDetailsForCrosswalksValidator checker;
+  const auto & issues = checker(*map);
+
+  EXPECT_EQ(issues.size(), 1);
+  EXPECT_EQ(issues[0].id, 24);
+  EXPECT_EQ(issues[0].severity, lanelet::validation::Severity::Error);
+  EXPECT_EQ(issues[0].primitive, lanelet::validation::Primitive::Polygon);
+  EXPECT_EQ(
+    issues[0].message,
+    "Crosswalk polygon of crosswalk regulatory element must have type of Crosswalk_polygon.");
```
Comment (YamatoAndo) [inline map/autoware_lanelet2_map_validator/test/src/test_regulatory_elements_details_for_crosswalks.cpp:82]:
```suggestion
    "Crosswalk polygon of crosswalk regulatory element must have type of crosswalk_polygon.");
```

And fix [here](https://github.com/autowarefoundation/autoware_tools/blob/2e93ca14722ff6ea97adf47f636e8614b3529e4a/map/autoware_lanelet2_map_validator/src/validators/crosswalk/regulatory_element_details_for_crosswalks.cpp#L122), too.

**Comment by TaikiYamada4** [comment]
@YamatoAndo 
Thanks for your review! All three suggestions above are fixed in here https://github.com/autowarefoundation/autoware_tools/pull/150/commits/2b3cfe81e15d59ff1c3227784b7f6d8b07e9ca21

**Comment by TaikiYamada4** [comment]
@YamatoAndo I've fixed the sample_map.osm `crosswalk_polygon` to `detection_area` so please check it out.

File: map/autoware_lanelet2_map_validator/test/src/map_validation_tester.hpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+#include <gtest/gtest.h>
+#include <lanelet2_io/Io.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+class MapValidationTester : public ::testing::Test
+{
+protected:
```
Comment (soblin) [inline map/autoware_lanelet2_map_validator/test/src/map_validation_tester.hpp:30]:
```suggestion
protected:
virtual void SetUp() {
  map_ = lanelet::load(package_share_directory + "/data/map/" + file_name, *projector, &loading_errors_);
```
}

Comment (TaikiYamada4) [reply map/autoware_lanelet2_map_validator/test/src/map_validation_tester.hpp:30]: Isn't this difficult to pass the target map file to validate?
One test code will try to validate various maps.

Comment (soblin) [reply map/autoware_lanelet2_map_validator/test/src/map_validation_tester.hpp:30]: Please forget my comment. Possibly we can use `TestWithParam` in the future when the number of test suites grow.
https://github.com/google/googletest/blob/main/docs/advanced.md#how-to-write-value-parameterized-tests

**Pull request merged**